File: src/renderer/canvas.zig
const std = @import("std");
const Color = @import("../nmath/color.zig").Color;

pub const Canvas = struct {
    width: usize,
    height: usize,
    buffer: []Color,

    pub fn init(allocator: std.mem.Allocator, width: usize, height: usize) !Canvas {
        const buffer_size = width * height;
        const buffer = try allocator.alloc(Color, buffer_size);

        for (buffer) |*p| {
            p.* = Color.init(0, 0, 0);
        }

        return .{
            .width = width,
            .height = height,
            .buffer = buffer,
        };
    }

    pub fn deinit(self: *Canvas, allocator: std.mem.Allocator) void {
        allocator.free(self.buffer);
    }

    pub fn pixelAt(self: Canvas, x: usize, y: usize) Color {
        return self.buffer[y * self.width + x];
    }

    pub fn writePixel(self: *Canvas, x: usize, y: usize, color: Color) void {
        self.buffer[y * self.width + x] = color;
    }

    fn constructPPMHeader(self: Canvas, allocator: std.mem.Allocator) ![]u8 {
        return try std.fmt.allocPrint(allocator, "P3\n{d} {d}\n255", .{ self.width, self.height });
    }

    fn constructPPMBody(self: Canvas, allocator: std.mem.Allocator) ![]u8 {
        var ppm_body: std.ArrayList(u8) = .empty;
        errdefer ppm_body.deinit(allocator);

        var width_counter: usize = 0;
        var digit_buffer: [4]u8 = undefined;

        var at_line_start = true;
        for (self.buffer) |c| {
            inline for (0..3) |j| {
                const color_channel: u8 = @intFromFloat(
                    std.math.clamp(std.math.round(c.vec.at(j) * 255), 0, 255),
                );
                const written_slice =
                    try std.fmt.bufPrint(&digit_buffer, "{d}", .{color_channel});

                const width_needed = written_slice.len + @intFromBool(!at_line_start);

                if (width_counter + width_needed > 70) {
                    try ppm_body.print(allocator, "\n", .{});
                    width_counter = 0;
                    at_line_start = true;
                }

                if (!at_line_start) {
                    try ppm_body.print(allocator, " ", .{});
                }

                try ppm_body.print(allocator, "{s}", .{written_slice});
                width_counter += width_needed;
                at_line_start = false;
            }
        }

        try ppm_body.print(allocator, "\n\n", .{});

        return ppm_body.toOwnedSlice(allocator);
    }

    pub fn toPPM(self: Canvas, allocator: std.mem.Allocator) ![]u8 {
        const ppm_header = try self.constructPPMHeader(allocator);
        const ppm_body = try self.constructPPMBody(allocator);
        return try std.fmt.allocPrint(allocator, "{s}\n{s}\n", .{ ppm_header, ppm_body });
    }
};

File: pkg/gfx/canvas.go
package gfx

import (
	"fmt"
	"math"
	"strings"

	. "github.com/novelalex/soft-raytracer/pkg/nmath"
	"github.com/novelalex/soft-raytracer/pkg/nutil"
)

type Canvas struct {
	width  uint
	height uint
	buffer []Color
}

func NewCanvas(width, height uint) Canvas {
	buffer_size := width * height
	buffer := make([]Color, buffer_size)
	return Canvas{width, height, buffer}
}

func (c Canvas) Width() uint {
	return c.width
}

func (c Canvas) Height() uint {
	return c.height
}

func (c Canvas) PixelAt(x, y uint) Color {
	return c.buffer[y*c.width+x]
}

func (c *Canvas) WritePixel(x, y uint, color Color) {
	c.buffer[y*c.width+x] = color
}

func (c Canvas) constructPPMHeader() string {
	return fmt.Sprintf("P3\n%d %d\n255\n", c.width, c.height)
}

func (c Canvas) constructP6PPMHeader() []byte {
	return []byte(fmt.Sprintf("P6\n%d %d\n255\n", c.width, c.height))
}

func (c Canvas) constructP6PPMBody() []byte {
	result := []byte{}
	for _, color := range c.buffer {
		for i := range 3 {
			cc := uint8(math.Max(0, math.Min(math.Round(color.At(i)*255), 255)))
			result = append(result, byte(cc))
		}
	}

	return result
}

func (c Canvas) AsP6PPM() []byte {
	result := []byte{}
	result = append(result, c.constructP6PPMHeader()...)
	result = append(result, c.constructP6PPMBody()...)
	return result
}

func (c Canvas) constructPPMBody() string {
	var sb strings.Builder
	width_counter := 0
	ln_start := true

	for _, color := range c.buffer {
		for i := range 3 {
			cc := uint8(math.Max(0, math.Min(math.Round(color.At(i)*255), 255)))

			written_digits := fmt.Sprintf("%d", cc)
			width_needed := len(written_digits) + nutil.IntFromBool(!ln_start)

			if width_counter+width_needed > 70 {
				sb.WriteString("\n")
				width_counter = 0
				ln_start = true
			}

			if !ln_start {
				sb.WriteString(" ")
			}

			sb.WriteString(written_digits)
			width_counter += width_needed
			ln_start = false
		}
	}
	sb.WriteString("\n\n")

	return sb.String()
}

func (c Canvas) AsPPM() string {
	ppm_header := c.constructPPMHeader()
	ppm_body := c.constructPPMBody()
	return ppm_header + ppm_body
}
